File: solutions/0200-number-of-islands/solution.py
class Solution(object):
    def numIslands(self, grid):
        """
        :type grid: List[List[str]]
        :rtype: int
        """
        dirs = [[1,0],[0,1],[-1,0],[0,-1]]
        rows = len(grid)
        columns = len(grid[0])
        islands = 0
        def bfs(r,c):
            q = deque()
            q.append((r,c))
            grid[r][c] = "0"
            while q:
                ro,co = q.popleft()
                for dr,dc in dirs:
                    nr,nc = ro+dr,co+dc
                    if (nr<0 or nc < 0 or nr>=rows or nc >= columns or grid[nr][nc]=="0"):
                        continue
                    q.append((nr,nc))
                    grid[nr][nc] = "0"

        
        for ro in range(rows):
            for co in range(columns):
                if grid[ro][co] == "1":
                    bfs(ro,co)
                    islands +=1
        return islands

        

File: solutions/0200-number-of-islands/solution.java
class Solution {

    public int numIslands(char[][] grid) {

        int m = grid.length;
        int n = grid[0].length;

        int ans = 0;

        for(int i = 0; i < m; i++){
            for(int j = 0; j < n; j++){

                if(grid[i][j] == '1'){
                    ans++;
                    bfs(i, j, grid);
                }
            }
        }

        return ans;
    }

    public void bfs(int i, int j, char[][] grid){

        int[][] dirs = {
            {1,0},
            {0,1},
            {-1,0},
            {0,-1}
        };

        grid[i][j] = '0';

        for(int[] d : dirs){

            int nr = i + d[0];
            int nc = j + d[1];

            if(nr < 0 || nc < 0 || nr >= grid.length || nc >= grid[0].length){
                continue;
            }

            if(grid[nr][nc] == '1'){
                bfs(nr, nc, grid);
            }
        }
    }
}
